apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,323 @@
+scenarios involving `vmap(vjp(fn))` or `vmap(jvp(fn))`.
+
+## Support for `chunk_size` in `vmap` and `jacrev`
+
+The computation of the Jacobian can be memory-intensive, and users have raised
+concerns about related issues, (e.g., [this one](https://github.com/pytorch/functorch/issues/680)).
+In response to these concerns, we have introduced a feature that allows for the
+calculation of `jacrev` and `vmap` in smaller, user-defined chunks, determined
+by the `chunk_size` argument. This adjustment serves to reduce the peak memory
+usage during the computation process. With this argument, users can specify the
+number of rows of the Jacobian to be computed at once. This enhancement was
+incorporated into the [jacrev PR](https://github.com/pytorch/pytorch/pull/89376)
+and the [vmap PR](https://github.com/pytorch/pytorch/pull/91157).
+
+## Support for `linearize` transform
+
+The "jvp" transform is designed to calculate both `f(x)` and the Jacobian-vector
+product. Consequently, even when one intends to compute the Jacobian-vector
+product for fixed inputs, the `jvp` transform still redundantly evaluates `f(x)`.
+To address such scenarios, the `linearize` transform comes into play. This
+transform proves valuable when multiple `jvp` computations are needed for
+constant inputs.
+
+Note that, in order to implement this efficiently, `linearize` stores some
+intermediate computations, which can result in higher memory requirements compared
+to directly applying `jvp`. The `linearize` transform was implemented in this
+[PR](https://github.com/pytorch/pytorch/pull/94173).
+
+## Support of `torch.func` transforms within `torch.compile`
+
+PyTorch 2.0 introduced a JIT compiler under `torch.compile`, similar to `jax.jit`.
+This opened up the possibility of compiling the existing transforms to enhance
+their performance. To understand how these transforms can be compiled, it is
+essential to discuss the workings of the three layers within the compilation
+stack, namely `dynamo`, `aot_autograd`, and `inductor`.
+
+The `dynamo` and `aot_autograd` layers primarily focus on capturing the
+computation graph and converting the captured operations into more basic operations.
+This captured graph is then passed to `inductor`, the compiler. `inductor` then applies
+various optimization passes before generating specialized code.
+
+To gain insight into the different stages of this stack,
+let us compile a simple program in debug mode.
+
+```python
+# Run this file with `TORCH_COMPILE_DEBUG=1`
+
+import torch
+
+def fn(x):
+    return torch.sin(x) + torch.square(x)
+
+torch.compile(fn)(torch.randn(4, 4))
+```
+
+**dynamo**: The primary responsibility of dynamo is to trace the Python program
+and convert it into the FX graph format. The FX graph generated by `dynamo`
+represents PyTorch operations from the public API, such as `torch.sin`.
+Below, you can observe the graph captured by `dynamo`.
```

Comment:
How are we observing a graph here? There's no figure. Does the reader have to install PyTorch and run this code in order to be shown a graph?

Reply:
Hmmm, interesting. I was expecting figures to illustrate the graphs.

But -- now I'm guessing: are these graphs autogenerated code? And thus there are no figures to be shown for the graphs, just the autogen code?

Reply:
yep, exactly. We call them graphs as they are generated by a tracer. This is standard in the compilers lingo. Feel free to add a small point in the post clarifying this.